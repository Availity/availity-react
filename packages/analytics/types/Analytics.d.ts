export interface AnalyticsEvent {
  [key: string]: any;
  url: string;
}

export interface AnalyticsPlugin {
  isEnabled: () => boolean;
  init: () => void;
  trackEvent: (event: AnalyticsEvent) => void;
  trackPageView: (url: string) => void;
}

export interface AnalyticsProps {
  children?: React.ReactNode;
  plugins?: AnalyticsPlugin[];
  pageTracking?: boolean;
  autoTrack?: boolean;
  recursive?: boolean;
  attributePrefix?: string;
  eventModifiers?: string | string[];
}

export interface TrackEventOptions {
  [key: string]: any;
  level: string;
}
export interface AnalyticsContext {
  trackEvent(trackEventOptions: TrackEventOptions): Promise<void>;
}

declare const Analytics: React.FunctionComponent<AnalyticsProps>;

declare const useAnalytics: () => AnalyticsContext;

export { useAnalytics };

export default Analytics;
