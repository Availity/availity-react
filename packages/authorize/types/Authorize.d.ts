import { QueryOptions } from 'react-query';

export type NumberOrString = string | number;
export type Permission = string | number | NumberOrString[];
export type Resource = string | number | NumberOrString[];

export interface AuthorizeProps {
  permissions: string | number | Permission[];
  resources?: string | number | Resource[];
  region?: boolean | string;
  loader?: boolean | React.ReactNode;
  organizationId?: string;
  customerId?: string;
  unauthorized?: React.ReactNode;
  negate?: boolean;
  queryOptions?: QueryOptions<boolean>;
}

declare const Authorize: React.FC<AuthorizeProps>;
export default Authorize;
