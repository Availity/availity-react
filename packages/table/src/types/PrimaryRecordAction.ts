export interface PrimaryRecordAction<T> {
  iconName: string | ((record?: T) => string);
  title: string | ((record?: T) => string);
  onClick: (record?: T) => void;
  isVisible?: (record?: T) => boolean;
}
