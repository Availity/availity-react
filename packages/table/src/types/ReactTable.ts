import {
  Row as RtRow,
  Cell as RtCell,
  HeaderGroup,
  Column as RtColumn,
  TableInstance as RtTableInstance,
  TableOptions as RtTableOptions,
  UseSortByInstanceProps,
  UseRowSelectInstanceProps,
  TableState,
  UseSortByColumnProps,
  UseTableHeaderGroupProps,
  UseRowSelectOptions,
  UseTableOptions,
  UseSortByOptions,
  UsePaginationOptions,
  UsePaginationInstanceProps,
  UseColumnOrderInstanceProps,
  UseRowSelectState,
} from 'react-table';

import { TableSort } from './TableSort';

/* eslint-disable-next-line  @typescript-eslint/ban-types */
export type IdType = { id?: string | number } & object;

export type Row<T extends IdType> = RtRow<T> & {
  toggleRowSelected: () => void;
  getToggleRowSelectedProps: () => void;
  original: T;
};

export type RowProps = {
  isDisabled?: boolean;
} & React.HTMLAttributes<HTMLTableRowElement>;

export type Column<T extends IdType> = RtColumn<T> & {
  className?: string;
  stickyRight?: boolean;
  stickyLeft?: boolean;
  disableClick?: boolean;
  title?: string;
  defaultCanSort?: boolean;
  disableSortBy?: boolean;
  label?: string;
  hidden?: boolean;
  isVisible?: boolean;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Cell<T extends IdType, V = any> = RtCell<T, V> & {
  index: number;
  row: Row<T>;
  column: Column<T>;
  original: T;
  data: T[];
};

export type ExtendedTableHeader<T extends IdType> = HeaderGroup<T> &
  Column<T> &
  UseSortByColumnProps<T> &
  UseTableHeaderGroupProps<T> & {
    defaultCanSort?: boolean;
    disableSortBy?: boolean;
    isSorted?: boolean;
    isSortedDesc?: boolean;
    toggleSortBy: (isDesc: boolean, isMult: boolean) => void;
  };

export type TableOptions<T extends IdType> = RtTableOptions<T> &
  UseTableOptions<T> & {
    autoResetSelectedRows?: boolean;
    autoResetSortBy?: boolean;
    manualSortBy?: boolean;
    hiddenColumns?: string[];
    initialState?: Partial<TableState<T>> & CurrentTableState<T>;
  } & UseSortByOptions<T> &
  UsePaginationOptions<T>;

export type TableInstance<T extends IdType> = UseSortByInstanceProps<T> &
  UseRowSelectInstanceProps<T> &
  UseRowSelectOptions<T> &
  RtTableInstance<T> &
  UseColumnOrderInstanceProps<T> &
  UsePaginationInstanceProps<T> & {
    toggleAllPageRowsSelected: (set?: boolean) => void;
    selectedFlatRows: Row<T>[];
    isAllRowsSelected: boolean;
    toggleAllRowsSelected: () => void;
    sortBy?: TableSort;
  } & {
    manualSortBy?: boolean;
  };

export type CurrentTableState<T extends IdType> = TableState &
  Partial<UseRowSelectState<T>> & {
    sortBy?: TableSort[];
  };
