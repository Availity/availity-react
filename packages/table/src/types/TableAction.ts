import React from 'react';

export interface TableAction {
  id: string;
  divider?: boolean;
}
