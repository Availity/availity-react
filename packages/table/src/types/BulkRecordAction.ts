import { TableAction } from './TableAction';

export interface BulkRecordAction<T> extends TableAction {
  onClick?: (records?: T[]) => void;
  isVisible?: (records?: T[]) => boolean;
  displayText?:
    | string
    | React.ReactChild
    | React.ElementType
    | ((records?: T[]) => string | React.ReactChild | React.ElementType);
}
