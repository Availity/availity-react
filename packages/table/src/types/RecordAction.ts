import { TableAction } from './TableAction';

export interface RecordAction<T> extends TableAction {
  onClick?: (record?: T) => void;
  isVisible?: (record?: T) => boolean;
  displayText?:
    | string
    | React.ReactChild
    | React.ElementType
    | ((record?: T) => string | React.ReactChild | React.ElementType);
}
