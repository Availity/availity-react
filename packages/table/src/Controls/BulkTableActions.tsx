import React, { useEffect, useState } from 'react';
import { Badge, Button, ButtonDropdown, ButtonGroup, DropdownItem, DropdownMenu, DropdownToggle } from 'reactstrap';
import { useTableContext } from '../TableContext';
import { BulkRecordAction } from '../types/BulkRecordAction';
import { IdType, TableInstance } from '../types/ReactTable';

type Props<T> = {
  id?: string;
  disabled?: boolean;
  recordName?: string;
  bulkActions?: BulkRecordAction<T>[];
  color?: string;
  onRecordsSelected?: (records: T[]) => void;
} & React.HTMLAttributes<HTMLElement>;

const BulkTableActions = <T extends IdType>({
  id,
  disabled,
  color,
  recordName = 'Records',
  bulkActions,
  onRecordsSelected,
}: Props<T>): JSX.Element | null => {
  const { instance } = useTableContext();
  const { isAllRowsSelected, toggleAllRowsSelected, selectedFlatRows } = instance as TableInstance<T>;

  const [isSelectionDropdownOpen, setIsSelectionDropdownOpen] = useState(false);
  const [numberOfSelectedRows, setNumberOfSelectedRows] = useState(0);
  const [selectionButtonText, setSelectionButtonText] = useState('Select');
  const [isDisabled, setIsDisabled] = useState(disabled || false);

  useEffect(() => {
    setNumberOfSelectedRows(selectedFlatRows?.length);
    if (isAllRowsSelected) {
      setSelectionButtonText('Deselect');
    } else {
      setSelectionButtonText('Select');
    }

    if (onRecordsSelected) {
      const records = selectedFlatRows.map((row: { original: T }) => row.original);
      onRecordsSelected(records);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedFlatRows]);

  useEffect(() => {
    setIsDisabled(disabled || false);
  }, [disabled]);

  const toggleSelectionDropdown = () => setIsSelectionDropdownOpen(!isSelectionDropdownOpen);

  const handleToggleSelectAll = () => {
    if (toggleAllRowsSelected) {
      toggleAllRowsSelected();
    }
  };

  return (
    <ButtonGroup data-testid="bulk_actions_btn_group" id={id} disabled={isDisabled} className="btn-group">
      <Button
        data-testid="select_deselect_all_records"
        disabled={disabled}
        onClick={() => {
          handleToggleSelectAll();
        }}
        color={color}
      >
        <Badge pill>{numberOfSelectedRows}</Badge> {selectionButtonText} All {recordName}
      </Button>
      {bulkActions && (
        <ButtonDropdown isOpen={isSelectionDropdownOpen} toggle={toggleSelectionDropdown}>
          <DropdownToggle
            data-testid="bulk_actions_toggle"
            disabled={numberOfSelectedRows === 0 || disabled}
            color={color}
            caret
          />
          <DropdownMenu color={color}>
            {bulkActions?.map((action) => {
              const isVisible = action.isVisible ? action.isVisible(selectedFlatRows.map((row) => row.original)) : true;
              const setProps = () => {
                if (!action.onClick) {
                  return null;
                }
                const clickEvent = action.onClick;
                return { onClick: () => clickEvent(selectedFlatRows.map((row) => row.original)) };
              };

              if (isVisible) {
                return action.divider ? (
                  <DropdownItem data-testid={`bulk_action_${action.id}`} key={action.id} divider />
                ) : (
                  <DropdownItem data-testid={`bulk_action_${action.id}`} key={`${action.id}`} {...setProps()}>
                    {action.displayText}
                  </DropdownItem>
                );
              }
              return null;
            })}
          </DropdownMenu>
        </ButtonDropdown>
      )}
    </ButtonGroup>
  );
};

export default BulkTableActions;
