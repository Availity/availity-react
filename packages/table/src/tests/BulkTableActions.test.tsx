import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import every from 'lodash/every';
import basicData from './data/basicData.json';
import TableControls from '../Controls/TableControls';
import BulkTableActions from '../Controls/BulkTableActions';
import { Column } from '../types/ReactTable';
import Table from '../Table';
import TableContent from '../TableContent';

const basicColumns = [
  {
    Header: 'First Name',
    accessor: 'first_name',
    defaultCanSort: true,
    disableSortBy: false,
  },
  {
    Header: 'Last Name',
    accessor: 'last_name',
    defaultCanSort: false,
    disableSortBy: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
  },
] as Column<Record<string, string | boolean | number>>[];

const bulkActions = [
  {
    id: 'action1',
    displayText: 'Action 1',
    isVisible: (records?: Record<string, string | boolean | number>[]) => every(records, { isActive: true }),
    onClick: jest.fn(),
  },
  {
    id: 'action2',
    displayText: 'Action 2',
    onClick: jest.fn(),
  },
  {
    id: 'action2',
    displayText: 'Action 2',
    onClick: jest.fn(),
    isVisible: () => false,
  },
  {
    id: 'action3',
    displayText: () => 'Populated By function',
  },
];

describe('BulkTableActions', () => {
  test('should render default bulk actions', async () => {
    const { container, getByTestId } = render(
      <Table data={basicData} columns={basicColumns}>
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
      </Table>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    expect(btnSelectDeselect.textContent).toBe('0 Select All Records');
  });

  test('should render bulk actions with record name', async () => {
    const { getByTestId } = render(
      <Table data={basicData} columns={basicColumns}>
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} recordName="MyRecords" />
        </TableControls>
      </Table>
    );

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    expect(btnSelectDeselect.textContent).toBe('0 Select All MyRecords');
  });

  test('should show number of selected records', async () => {
    const { getByTestId } = render(
      <Table
        initialState={{
          selectedRowIds: { '1': true },
        }}
        data={basicData}
        columns={basicColumns}
      >
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
        <TableContent />
      </Table>
    );

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    expect(btnSelectDeselect.textContent).toBe('1 Select All Records');
  });

  test('should show Deselect all rows when all rows are selected', async () => {
    const { getByTestId } = render(
      <Table
        initialState={{
          selectedRowIds: { '0': true, '1': true, '2': true },
        }}
        data={basicData}
        columns={basicColumns}
        paged={false}
      >
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
        <TableContent />
      </Table>
    );

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    expect(btnSelectDeselect.textContent).toBe('3 Deselect All Records');
  });

  test('should call bulk action when clicking event', async () => {
    const { getByTestId } = render(
      <Table
        initialState={{
          selectedRowIds: { '1': true },
        }}
        data={basicData}
        columns={basicColumns}
      >
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
        <TableContent />
      </Table>
    );

    const btnToggle = await waitFor(() => getByTestId('bulk_actions_toggle'));
    expect(btnToggle).not.toBeNull();
    fireEvent.click(btnToggle);

    const actionToDo = bulkActions[1];

    const btnAction = await waitFor(() => getByTestId(`bulk_action_${actionToDo.id}`));
    expect(btnAction).not.toBeNull();

    fireEvent.click(btnAction);

    await waitFor(() => {
      expect(actionToDo.onClick).toHaveBeenCalledTimes(1);
    });
  });

  test('should fire onRecordsSelected when provided', async () => {
    const onRowsSelected = jest.fn();

    const { getByTestId } = render(
      <Table
        initialState={{
          selectedRowIds: { '1': true },
        }}
        data={basicData}
        columns={basicColumns}
      >
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} onRecordsSelected={onRowsSelected} />
        </TableControls>
        <TableContent />
      </Table>
    );

    const btnSelectDeselect = await waitFor(() => getByTestId('select_deselect_all_records'));
    expect(btnSelectDeselect).not.toBeNull();

    fireEvent.click(btnSelectDeselect);

    await waitFor(() => {
      expect(onRowsSelected).toHaveBeenCalled();
    });
  });
});
