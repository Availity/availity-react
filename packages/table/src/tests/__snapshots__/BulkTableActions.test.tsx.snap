// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkTableActions should render default bulk actions 1`] = `
<div>
  <div
    class="table-controls"
    data-testid="test-controls-container"
  >
    <form
      class="form-inline"
    >
      <div
        class="btn-group btn-group"
        data-testid="bulk_actions_btn_group"
        role="group"
      >
        <button
          class="btn btn-secondary"
          data-testid="select_deselect_all_records"
          type="button"
        >
          <span
            class="badge badge-secondary badge-pill"
          >
            0
          </span>
           
          Select
           All 
          Records
        </button>
        <div
          class="btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary disabled"
            data-testid="bulk_actions_toggle"
            disabled=""
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="bulk_action_action1"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 1
            </button>
            <button
              class="dropdown-item"
              data-testid="bulk_action_action2"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 2
            </button>
            <button
              class="dropdown-item"
              data-testid="bulk_action_action3"
              role="menuitem"
              tabindex="0"
              type="button"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
